lib/linter/code-path-analysis/code-path-segment.js

```diff
@@ -196,19 +225,21 @@ class CodePathSegment {
     }
 
     /**
-     * Replaces unused segments with the previous segments of each unused segment.
-     * @param {CodePathSegment[]} segments An array of segments to replace.
-     * @returns {CodePathSegment[]} The replaced array.
+     * Creates a new array based on an array of segments. If any segment in the
+     * array is unused, then it is replaced by all of its previous segments.
+     * All used segments are returned as-is without replacement.
+     * @param {CodePathSegment[]} segments The array of segments to flatten.
+     * @returns {CodePathSegment[]} The flattened array.
      */
     static flattenUnusedSegments(segments) {
-        const done = Object.create(null);
+        const done = new Set();
         const retv = [];
```

Comment:
Now that `retv` merely replicates the contents of `done`, I think that we could remove it completely and just return the contents of `done` as an array instead, i.e. (at the end of the function):

```js
return [...done];
```

Reply:
I'm not sure that sets guarantee the order of the items that are added to it, which is why I kept the array.

Reply:
The order of iteration of elements in a `Set` is defined to be the same as the order of insertion. This means that `[...done]` has elements in the same order as they were `add()`ed to the set - when not already included.

This is explained for example in [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#set.prototypeiterator) where it states that `Set.prototype[Symbol.iterator]()`:

> Returns a new iterator object that yields the **values** for each element in the `Set` object in insertion order.

Ultimately, this behavior is defined by the specification, although in a more technical and abstract language. A `Set` is specified as having a list (the internal property [_[[SetData]]_](https://262.ecma-international.org/14.0/#_ref_12795)) where element are stored in an ordered sequence, so the order of the elements in reading and writing operations is well-defined and does not depend on the implementation.

Reply:
Ah nice, thanks for the reference. I'll update the code.